File: project/backend/app/models/organization.py
"""
Organization model for healthcare provider organisations
"""
from datetime import datetime

from sqlalchemy import Column, DateTime, Float, Integer, Numeric, String, Text
from sqlalchemy.orm import relationship

from app.db.base import Base


class Organization(Base):
    """
    Represents a healthcare organisation (hospital, clinic, practice).

    Revenue and utilisation are analytics fields pre-populated from external
    datasets; lat/lon enable map-based lookups.  This is a read-mostly table
    managed outside the patient-care workflow.
    """
    __tablename__ = "organizations"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(255), nullable=False, index=True)
    address = Column(Text, nullable=True)
    city = Column(String(100), nullable=True, index=True)
    state = Column(String(50), nullable=True)
    zip = Column(String(20), nullable=True)
    phone = Column(String(30), nullable=True)
    revenue = Column(Numeric(15, 2), nullable=True)
    utilization = Column(Float, nullable=True)
    lat = Column(Float, nullable=True)
    lon = Column(Float, nullable=True)
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)

    # Providers affiliated with this organisation
    providers = relationship("Provider", back_populates="organization")

    def __repr__(self) -> str:
        return f"<Organization(id={self.id}, name='{self.name}')>"

File: project/mobile/lib/models/organization.dart
/// Organization model representing a healthcare organisation.
class Organization {
  final int id;
  final String name;
  final String? address;
  final String? city;
  final String? state;
  final String? zip;
  final String? phone;
  final double? revenue;
  final double? utilization;
  final double? lat;
  final double? lon;
  final DateTime createdAt;

  Organization({
    required this.id,
    required this.name,
    this.address,
    this.city,
    this.state,
    this.zip,
    this.phone,
    this.revenue,
    this.utilization,
    this.lat,
    this.lon,
    required this.createdAt,
  });

  factory Organization.fromJson(Map<String, dynamic> json) {
    return Organization(
      id: json['id'] as int,
      name: json['name'] as String,
      address: json['address'] as String?,
      city: json['city'] as String?,
      state: json['state'] as String?,
      zip: json['zip'] as String?,
      phone: json['phone'] as String?,
      revenue: (json['revenue'] as num?)?.toDouble(),
      utilization: (json['utilization'] as num?)?.toDouble(),
      lat: (json['lat'] as num?)?.toDouble(),
      lon: (json['lon'] as num?)?.toDouble(),
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  /// Formatted city/state string for list display.
  String get location {
    if (city != null && state != null) return '$city, $state';
    if (city != null) return city!;
    if (state != null) return state!;
    return 'Unknown';
  }

  /// Utilization as a percentage string, e.g. "78.4%".
  String get utilizationLabel =>
      utilization != null ? '${(utilization! * 100).toStringAsFixed(1)}%' : 'N/A';
}
